js/ui/date_box/ui.date_box.mask.js

```diff
@@ -124,30 +125,82 @@ const DateBoxMask = DateBoxBase.inherit({
     },
 
     _isSingleCharKey(e) {
-        const key = e.originalEvent.key;
+        const key = e.originalEvent.data || e.originalEvent.key;
         return typeof key === 'string' && key.length === 1 && !e.ctrl && !e.alt;
     },
 
+    _useBeforeInputEvent: function() {
+        const device = devices.real();
+        return device.android && device.version[0] > 4;
+    },
+
     _keyboardHandler(e) {
         const key = e.originalEvent.key;
 
         const result = this.callBase(e);
 
-        if(!this._useMaskBehavior() || !this._isSingleCharKey(e)) {
+        if(!this._useMaskBehavior() || this._useBeforeInputEvent() || !this._isSingleCharKey(e)) {
             return result;
         }
 
-        if(this._isAllSelected()) {
-            this._activePartIndex = 0;
+        this._processInputKey(key);
+
+        e.originalEvent.preventDefault();
+
+        return result;
+    },
+
+    _maskBeforeInputHandler(e) {
+        this._maskInputHandler = null;
+
+        const { inputType } = e.originalEvent;
+
+        if(inputType === 'insertCompositionText') {
+            this._maskInputHandler = () => {
+                this._renderDisplayText(this._getDisplayedText(this._maskValue));
+                this._selectNextPart();
+            };
         }
 
-        this._setNewDateIfEmpty();
+        const isBackwardDeletion = inputType === 'deleteContentBackward';
+        const isForwardDeletion = inputType === 'deleteContentForward';
+        if(isBackwardDeletion || isForwardDeletion) {
+            const direction = isBackwardDeletion ? BACKWARD : FORWARD;
+            this._maskInputHandler = () => {
+                this._revertPart();
+                this._selectNextPart(direction);
+            };
+        }
+
+        if(!this._useMaskBehavior() || !this._isSingleCharKey(e)) {
+            return;
+        }
 
-        isNaN(parseInt(key)) ? this._searchString(key) : this._searchNumber(key);
+        const key = e.originalEvent.data;
+        this._processInputKey(key);
+        e.preventDefault();
+        return true;
+    },
 
-        e.originalEvent.preventDefault();
+    _keyPressHandler(e) {
+        this.callBase(e);
 
-        return result;
+        if(this._maskInputHandler) {
+            this._maskInputHandler();
+            this._maskInputHandler = null;
+        }
+    },
+
+    _processInputKey(key) {
+        if(this._isAllSelected()) {
+            this._activePartIndex = 0;
+        }
+        this._setNewDateIfEmpty();
+        if(isNaN(parseInt(key))) {
```

Comment:
It will be false on "123someLetters". Is it ok?

Reply:
The `key` argument is always one char - we have `!this._isSingleCharKey(e)` conditions before the `_processInputKey` call:
https://github.com/DevExpress/DevExtreme/pull/12355/files#diff-e23fb8153c04867cc50fe7f201618494R142
https://github.com/DevExpress/DevExtreme/pull/12355/files#diff-e23fb8153c04867cc50fe7f201618494R175